src/yunohost/certificate.py

```diff
@@ -0,0 +1,641 @@
+    # Let's Encrypt, great !
+    elif (CAtype == "Let's Encrypt") :                                   statusSummaryCode =  20
+
+    return { "domain"      : domain,
+             "subject"     : certSubject,
+             "CAname"      : certIssuer,
+             "CAtype"      : CAtype,
+             "validity"    : daysRemaining,
+             "summaryCode" : statusSummaryCode
+           }
+
+###############################################################################
+#   Misc small stuff ...                                                      #
+###############################################################################
+
+def _generate_account_key() :
+
+    logger.info("Generating account key ...")
+    _generate_key(account_key_file)
+    _set_permissions(account_key_file, "root", "root", 0400)
+
+def _generate_key(destinationPath) :
+
+    k = crypto.PKey()
+    k.generate_key(crypto.TYPE_RSA, key_size)
+
+    with open(destinationPath, "w") as f :
+        f.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, k))
+
+def _set_permissions(path, user, group, permissions) :
+
+    uid = pwd.getpwnam(user).pw_uid
+    gid = grp.getgrnam(group).gr_gid
+
+    os.chown(path, uid, gid)
+    os.chmod(path, permissions)
+
+def _backup_current_cert(domain):
+
+    logger.info("Backuping existing certificate for domain "+domain)
+
+    cert_folder_domain = cert_folder+"/"+domain
+
+    dateTag = datetime.now().strftime("%Y%m%d.%H%M%S")
+    backup_folder = cert_folder_domain+"-backup-"+dateTag
+
+    shutil.copytree(cert_folder_domain, backup_folder)
+
+
+def _check_domain_is_correctly_configured(domain) :
+
+    public_ip = yunohost.domain.get_public_ip()
+
+    # Check if IP from DNS matches public IP
+    if not _dns_ip_match_public_ip(public_ip, domain) :
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_domain_dns_ip_differs_from_public_ip', domain=domain))
+
+    # Check if domain seems to be accessible through HTTP ?
+    if not _domain_is_accessible_through_HTTP(public_ip, domain) :
+        raise MoulinetteError(errno.EINVAL,  m18n.n('certmanager_domain_http_not_working', domain=domain))
+        
+def _dns_ip_match_public_ip(public_ip, domain) :
+
+    try :
+        r = requests.get("http://dns-api.org/A/"+domain)
+    except :
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_error_contacting_dns_api', api="dns-api.org"))
+        
+    if (r.text == "[{\"error\":\"NXDOMAIN\"}]") :
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_no_A_dns_record', domain=domain))
+
+    try :
+        dns_ip = json.loads(r.text)[0]["value"]
+    except :
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_error_parsing_dns', domain=domain, value=r.text))
+
+    if (dns_ip != public_ip) :
+        return False
+    else :
+        return True
+
+def _domain_is_accessible_through_HTTP(ip, domain) :
```

Comment:
This test looks really weak, I don't think that this is the good way to do that but I don't have any other in mind right now.